raft: expose partition/raft state via admin API

bharathv opened the pull request:
Implements `GET /v1/debug/partition/{namespace}/{topic}/{partition}` to fetch a detailed partition debug json that includes internal state at the partition (cluster) and raft layer.  Can be queried on any a node. Handling node aggregates the output from all replicas of a given partition (including learners, if a reconfiguration is in progress).

Fixes #9097

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v23.1.x
- [ ] v22.3.x
- [ ] v22.2.x

## Release Notes

### Improvements

* Adds an admin API that provides insight into partition/raft state of all replicas of a partition.

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

bharathv commented:
Sample output from a 7 replica partition via `curl http://localhost:9644/v1/debug/partition/kafka/foo/0`

```
{
	"ntp": "{kafka/foo/0}",
	"replicas": [{
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 0,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 1,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": 1,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": 0,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": true,
			"is_elected_leader": true,
			"followers": [{
				"id": 6,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}, {
				"id": 3,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}, {
				"id": 0,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}, {
				"id": 4,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}, {
				"id": 5,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}, {
				"id": 2,
				"last_flushed_log_index": 0,
				"last_dirty_log_index": 0,
				"match_index": 0,
				"next_index": 1,
				"last_sent_offset": 0,
				"heartbeats_failed": 0,
				"is_learner": false,
				"ms_since_last_heartbeat": 135,
				"last_sent_seq": 112736,
				"last_received_seq": 112735,
				"last_successful_received_seq": 112735,
				"suppress_heartbeats": false,
				"is_recovering": false
			}]
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 2,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 3,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 4,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 5,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}, {
		"start_offset": 0,
		"committed_offset": 0,
		"last_stable_offset": 1,
		"high_watermark": 1,
		"dirty_offset": 0,
		"latest_configuration_offset": 0,
		"revision_id": 45,
		"log_size_bytes": 173,
		"non_log_disk_size_bytes": 67,
		"is_read_replica_mode_enabled": false,
		"read_replica_bucket": "",
		"is_remote_fetch_enabled": false,
		"is_cloud_data_available": false,
		"start_cloud_offset": -1,
		"next_cloud_offset": -1,
		"raft_state": {
			"node_id": 6,
			"term": 1,
			"group_configuration": "{current: {voters: {{id: {1}, revision: {45}}, {id: {6}, revision: {45}}, {id: {3}, revision: {45}}, {id: {0}, revision: {45}}, {id: {4}, revision: {45}}, {id: {5}, revision: {45}}, {id: {2}, revision: {45}}}, learners: {}}, old:{nullopt}, revision: 45, update: {nullopt}, version: 5}}",
			"confirmed_term": -9223372036854775808,
			"flushed_offset": 0,
			"commit_index": 0,
			"majority_replicated_index": 0,
			"visibility_upper_bound_index": 0,
			"last_quorum_replicated_index": -9223372036854775808,
			"last_snapshot_term": -9223372036854775808,
			"received_snapshot_bytes": 0,
			"last_snapshot_index": -9223372036854775808,
			"received_snapshot_index": -9223372036854775808,
			"has_pending_flushes": false,
			"is_leader": false,
			"is_elected_leader": false
		}
	}]
}
```

bharathv commented:
Took a different approach than previous attempts to address this issue, differences are the following

* Unified single handler for partition state, rather than having separate partition/raft APIs
* Aggregated output in the same RPC for all replicas, rather than querying node local/leader replicas. A leader replica (if one exists) includes follower metadata.
* Can transparently handle one or more nodes hosting replicas crash (caller doesn't have to know which node to query).

mmaslankaprv commented on src/v/cluster/types.h:
```diff
@@ -3057,12 +3198,14 @@ struct partition_state_reply
       = default;
 
-    auto serde_fields() { return std::tie(ntp); }
+    auto serde_fields() { return std::tie(ntp, state); }
```
missing `error_code` ?

mmaslankaprv commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1395,4 +1397,100 @@ ss::future<std::error_code> topics_frontend::move_partition_replicas(
+    std::vector<partition_state> results;
+    results.reserve(finished_futures.size());
+    for (auto& fut : finished_futures) {
+        if (fut.failed()) {
```
this will result in ignored failed future as we do not get the exception

mmaslankaprv replied:
maybe worth logging an error ?

dotnwat commented on src/v/cluster/types.h:
```diff
@@ -3049,6 +3049,148 @@ struct partition_state_request
+      envelope<partition_state, serde::version<0>, serde::compat_version<0>> {
+    using rpc_adl_exempt = std::true_type;
+
+    model::offset _start_offset;
```
curious why all these are `_` prefixed?

dotnwat commented on tests/rptest/tests/partition_state_api_test.py:
```diff
@@ -0,0 +1,107 @@
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs, num_brokers=5)
+
+    def _validate_states(self, states, num_replicas, has_leader=True):
```
do you think it is worth sanity checking more of the result, or even checking the schema? might be overkill.

bharathv replied:
May be an overkill, this is a v0 of a debug endpoint, pretty sure this will change as we add more fields as needed for debugging incidents. (eg: stm specific items in the state)

dotnwat commented on src/v/cluster/topics_frontend.cc:
```diff
@@ -1391,4 +1397,118 @@ ss::future<std::error_code> topics_frontend::move_partition_replicas(
+              clusterlog.debug,
+              "Failed to get partition state for ntp: {}, failure: {}",
+              ntp,
+              fut.get_exception());
```
it's a minor thing, but can you pull this out of the log statement like:

```
auto ex = fut.get_exception();
vlog(..., ex);
```

The reason is that `get_exception` has an important side-effect (not throwing abandoned exceptional future) and I think it better to be a bit clearer to anyone in the future who might refactor something.

bharathv replied:
Done.

dotnwat reviewed (commented):
lgtm!

bharathv commented:
[Force push](https://github.com/redpanda-data/redpanda/compare/e9c8c9204b04ad0ed1e1173ca3803a728885a266..23012dd3b679cd2449e5b85267811e84606a7a28) addresses comments and a rebase to fix conflicts.

bharathv commented:
Failure: #8753 (unrelated, known issue).

dotnwat reviewed (approved):
epic

bharathv commented:
/backport  v23.1.x

vbotbuildovich commented:
Failed to run cherry-pick command. I executed the below command:

```
git cherry-pick -x 62c1e088cfcf5ad93e97639dffc5629833a24c01 95b8831a9848ce89c9a54d37eed0b7ec2ae87b6d 141fc77a5034182094deb859106998d7e404f9cd f6454a88f7277d80bc955a26bd7619d484d86ceb cc2a56a18f1b809ae1945663a36b7698af12eab6 3a565fa5ad375b2be20b239c8c9e90f405b6e2aa b8479fe14a282c6083964b15803fc61b3478965e b23da1d8307180ca53ecaef0013ecf75fd0c2bcb 23012dd3b679cd2449e5b85267811e84606a7a28
```

[Workflow run logs](https://github.com/redpanda-data/redpanda/actions/runs/4737437183).
